### PR TITLE
Add: `MultiFab::sum_unique`

### PR DESCRIPTION
**Pull request by ax3l**
## Summary

This provides a new method to sum values in a `MultiFab`.
For non-cell-centered data, `MultiFab::sum` double counts box boundary values that are owned by multiple boxes. This provides a function that does not double count these and provides a quick way to get only the sum of physically unique values.

- [ ] test: is there a good AMReX location to call this in an existing test (or write a new one)? I'll use this in ImpactX (see below)

## Additional background

https://github.com/ECP-WarpX/impactx/pull/165

## Checklist

The proposed changes:
- [ ] fix a bug or incorrect behavior in AMReX
- [x] add new capabilities to AMReX
- [ ] changes answers in the test suite to more than roundoff level
- [ ] are likely to significantly affect the results of downstream AMReX users
- [ ] include documentation in the code and/or rst files, if appropriate

### CODE REVIEW

File: Src/Base/AMReX_MultiFab.cpp [hunk 1]
```diff
@@ -1586,6 +1587,53 @@ MultiFab::sum (int comp, bool local) const
...
+    using namespace amrex::literals;
+
+    // Owner is the grid with the lowest grid number containing the data
+    std::unique_ptr<iMultiFab> owner_mask = OwnerMask(period);
+
+    Real sm = Real(0.0);
+#ifdef AMREX_USE_GPU
+    if (Gpu::inLaunchRegion()) {
+        auto const& ma = this->const_arrays();
+        auto const& msk = owner_mask->const_array();
```
Comment (ax3l) [inline Src/Base/AMReX_MultiFab.cpp]: @WeiqunZhang I am not sure how to pass the owner mast onto GPU `ParReduce` without an `MFIter`. What am I missing?

Comment (WeiqunZhang) [inline Src/Base/AMReX_MultiFab.cpp]:
```suggestion
        auto const& msk = owner_mask->const_arrays();
```

File: Src/Base/AMReX_MultiFab.cpp [hunk 2]
```diff
@@ -1586,6 +1587,53 @@ MultiFab::sum (int comp, bool local) const
...
+        sm = ParReduce(TypeList<ReduceOpSum>{}, TypeList<Real>{}, *this, IntVect(0),
+        [=] AMREX_GPU_DEVICE (int box_no, int i, int j, int k) noexcept
+                       -> GpuTuple<Real>
+        {
+            return msk(i,j,k,comp) ? ma[box_no](i,j,k,comp) : 0.0_rt;
```
Comment (WeiqunZhang) [inline Src/Base/AMReX_MultiFab.cpp]:
```suggestion
            return msk[box_no](i,j,k) ? ma[box_no](i,j,k,comp) : 0.0_rt;
```

File: Src/Base/AMReX_MultiFab.cpp [hunk 3]
```diff
@@ -1586,6 +1587,53 @@ MultiFab::sum (int comp, bool local) const
...
+#endif
+        for (MFIter mfi(*this,true); mfi.isValid(); ++mfi)
+        {
+            Box const& bx = mfi.tilebox();
+            Array4<Real const> const& a = this->const_array(mfi);
+            Array4<int const> const& msk = owner_mask->const_array(mfi);
+            Real tmp = 0.0_rt;
+            AMREX_LOOP_3D(bx, i, j, k,
+            {
+                tmp += msk(i,j,k,comp) ? a(i,j,k,comp) : 0.0_rt;
```
Comment (WeiqunZhang) [inline Src/Base/AMReX_MultiFab.cpp]:
```suggestion
                tmp += msk(i,j,k) ? a(i,j,k,comp) : 0.0_rt;
```

File: Src/Base/AMReX_MultiFab.cpp [hunk 4]
```diff
@@ -1586,6 +1587,53 @@ MultiFab::sum (int comp, bool local) const
...
 }
 
+Real
+MultiFab::sum_unique (int comp,
+                      const Periodicity& period) const
+{
+    BL_PROFILE("MultiFab::sum_unique()");
+
+    using namespace amrex::literals;
+
```
Comment (WeiqunZhang) [inline Src/Base/AMReX_MultiFab.cpp]:
```suggestion
```
The namespace is imported already.

**Review by WeiqunZhang — COMMENTED** [review]
I think we should return MultiFab::sum if it's cell centered.  That is we assume there are no overlaps for cell-centered multifabs, which is not always true, but true for almost all cases.

We could also add optional `bool local=false` to give the user a chance to skip MPI communication, because they might want to aggregate the communication.

**Comment by ax3l** [comment]
Thanks a lot for the review. Updated accordingly :)

**Review by WeiqunZhang — APPROVED** [review]
